std/algorithm/searching.d

```diff
@@ -1311,33 +1311,64 @@ if (isInputRange!Range && !isInfinite!Range &&
     alias CommonElement = CommonType!(Element, RangeElementType);
     Unqual!CommonElement extremeElement = seedElement;
 
-    alias MapType = Unqual!(typeof(mapFun(CommonElement.init)));
-    MapType extremeElementMapped = mapFun(extremeElement);
 
-    // direct access via a random access range is faster
-    static if (isRandomAccessRange!Range)
+    // if we only have one statement in the loop, it can be optimized a lot better
+    static if (__traits(isSame, map, a => a))
     {
-        foreach (const i; 0 .. r.length)
+
+        // direct access via a random access range is faster
```

Comment:
Is this really true? Looping over an index is faster than repeated `empty`/`front`/`popFront()`?

Reply:
Yes, D ranges have an overhead of ~20-50% for arrays. It's pretty bad for e.g. joiner (see https://gist.github.com/wilzbach/ffd5d20639766a831fd4b1962572897a), but it's a problem here too.

Check also https://github.com/dlang/phobos/pull/4265

The reason is that the compiler can inline the commands a lot better for native arrays and even do vectorization better than for the generic loop.